pycloudlib/openstack/cloud.py

```diff
@@ -0,0 +1,159 @@
+""" Openstack cloud type."""
+import base64
+from operator import itemgetter
+
+import openstack
+
+from pycloudlib.cloud import BaseCloud
+from pycloudlib.openstack.instance import OpenstackInstance
+
+
+class Openstack(BaseCloud):
+    """ Openstack cloud class."""
+
+    _type = 'openstack'
+
+    def __init__(self, tag, network, timestamp_suffix=True):
+        super().__init__(tag, timestamp_suffix)
+        self.network = network
+        self._openstack_keypair = None
+        self.conn = openstack.connect()
+
+    def delete_image(self, image_id):
+        self.conn.delete_image(image_id, wait=True)
+
+    def released_image(self, release, **kwargs):
+        """Find the id of the latest released image for a particular release.
+
+        Args:
+            release: string, Ubuntu release to look for
+            arch: string, architecture to use
+            root_store: string, root store to use
+
+        Returns:
+            string, id of latest image
+
+        """
+        self._log.warning('Released images NOT supported for openstack')
+        return self.daily_image(release)
+
+    def daily_image(self, release, **kwargs):
```

Comment:
I initially tried to use simplestreams here, but realized that many openstack deployments (including ours) are probably out of date with the latest images.

Reply:
Yeah, that's definitely an issue we'll hit with OpenStack deployments. It's also entirely possible, of course, that there are no Ubuntu images available in any given deployment.

I'm concerned that this heuristic could cause some confusing test behaviours, both between platforms (i.e. a current EC2 image vs a stale OpenStack image) and between OpenStack deployments (most likely one which is running glance-simplestreams-sync vs one which is not). We could also very easily end up launching non-base non-official images without noticing.

I wonder if instead of trying to determine this automatically, we should accept that we can't reliably launch the images that pycloudlib usually does and require OpenStack consumers to provide an explicit image ID in all cases? My thinking is that the image ID to use is deployment-specific, so is best determined by the consumers who know what OpenStack deployment they're running against.

(IIRC, what we did in cloud-test-framework is have an OpenStack cloud class which holds the cross-deployment behaviour and logic, and then subclasses of that which define the deployment-specific behaviour: ScalingStack and then the OpenStack public clouds we supported. Someone trying to use "generic OpenStack" would have to fill in the unknowns to proceed, but we still had a shared understanding of how to operate against shared deployments stored in the codebase.)

A couple of other thoughts: I believe that glance-simplestreams-sync does include additional metadata against the images, which we might be able to use to make this heuristic less likely to hit false positives (i.e. less likely to return non-official Ubuntu images); we might even be able to use it to cross-reference streams (i.e. gather a list of all fingerprints that are valid for $series according to streams and then find the most recent corresponding image). We could also consider using streams to populate the target deployment with an image ourselves: that's probably the only way we can be sure that we'll have an official Ubuntu image on every OpenStack deployment we encounter.